### ISSUE TITLE
NumberFormat bug when parsing percentages

### ISSUE THREAD

**Issue by aepshteyn** [gwtproject/gwt#9611]
**GWT version:** 2.8.2
**Browser (with version):** JUnitShell
**Operating System:** Windows 8.1

---

##### Description

`com.google.gwt.i18n.client.NumberFormat` fails to divide by 100 when parsing percentages.

##### Example
```
NumberFormat fmt = NumberFormat.getFormat("#.##%");
fmt.format(.1234);  // returns "12.34%" as expected
fmt.parse("12.34%");  // returns 12.34 (wtf?)
```
In the above example `fmt.parse("12.34%")` should obviously return `.1234` instead of `12.34`.

The `format()` and `parse()` methods should be symmetric as much as possible.  If `format(X)` multiplies X by 100, `parse("X%")` should divide X by 100.

##### Example from JRE
`java.text.DecimalFormat` does this properly:
```
DecimalFormat fmt = new DecimalFormat("#.##%");
String str = fmt.format(.1234); // returns "12.34%" as expected
fmt.parse(str);  // returns 0.1234 (as expected)
``

**Comment by niloc132** [comment]
Good catch - looks liker we're just missing a step at the end of parse to divide the result by `multiplier`, plus appropriate tests.

**Comment by jnehlmeier** [comment]
Hmm,

https://github.com/gwtproject/gwt/blob/master/user/src/com/google/gwt/i18n/client/NumberFormat.java#L1597 

does store a scale on parsing numbers and devides through it at the end.

**Comment by niloc132** [comment]
Good point - seems like the format string should be respected though? Worth thinking about.

Anyway, while `scale` is set, it isn't used if Double.parseDouble succeeds at the end - the `ret = ret / scale;` line needs to also be used in the try/catch. And tests added - there are nearly no parse tests at all.

**Comment by vegegoku** [comment]
Debugging a little bit and i found that the reason of this issue the call 
https://github.com/gwtproject/gwt/blob/fa41a03a93b74ead4ef2f93fff6c1d5bc9ede47f/user/src/com/google/gwt/i18n/client/NumberFormat.java#L1046

which at some point calls the `parseNumber` with the value only 
https://github.com/gwtproject/gwt/blob/fa41a03a93b74ead4ef2f93fff6c1d5bc9ede47f/user/src/com/google/gwt/i18n/client/NumberFormat.java#L1098
removing the the last `%` preventing `parseNumber` from calculating the correct scale leaving it to default value `1`

**Comment by vegegoku** [comment]
I have noticed from the following test :
```java
value = numberParse("0.0;(0.0)", "123.4579%");
assertTrue(value.doubleValue() == (123.4579 / 100));
```
that for percent there might be pattern that does not have `%`, in this case the scale will be calculated in the `parseNumber` from the string being parsed, and parsing such pattern will leave multiplier field value to `1`, while for pattern such as `#.##%` the pattern will be parsed and produce a multiplier value of `100` then the `%` will be excluded from the string being passed to `parseNumber` to be parsed so the scale wont be calculated and will remain on default value of `1`.

so as summary for parsing the same String `12.34%`

| Pattern        | Scale           | Multiplier  |
| ------------- |-------------:| -----:|
| 0.0;(0.0)      | 100 | 1 |
| #.##%      | 1      |   100 |

the cause of the issue for the `#.##%` pattern is that the scale is not being calculated, but also does not use the calculated multiplier.

this could be fixed by deviding the result of the `parseNumber` by the mulitplier here
https://github.com/gwtproject/gwt/blob/0c01db9b3992f42e04333ec6a04d4c1c9c3780ad/user/src/com/google/gwt/i18n/client/NumberFormat.java#L1098

so this will be `ret = parseNumber(valueOnly, tempPos) / multiplier;`

i added just one simple test case for the `#.##%` and it worked with the old tests

any suggestions or hints or should i make a patch with the fix?

**Comment by vegegoku** [comment]
according to the comment above, should i proceed with this on gerrit?

**Comment by niloc132** [comment]
Yes, if you have a patch and tests, feel free to put it on gerrit for review.

I think I would go further and remove `scale` from parseNumber, since it doesn't appear to make any sense to compute there, for two reasons: the token has already been stripped off, and that makes the scale based on the input string, rather than on the pattern (i.e. the intent of the rest of the class isn't to treat the percent symbol like an equivalent of E-2). The confusing and unused  `ret = ret / scale;` can then also be removed.

**Comment by vegegoku** [comment]
i will be fixing the tests too, and if there is any specific tests i need to add i am all ears and i will be glad to add them

**Comment by vegegoku** [comment]
I agree with @niloc132  here, the scale logic seems to be broken and is not aligned with the JRE logic
 
JRE

```java
DecimalFormat fmt = new DecimalFormat("#.##");
fmt.parse("123.45%"); // result : 123.45

DecimalFormat fmt = new DecimalFormat("#.##%");
fmt.parse("123.45%"); // result : 1.2345

DecimalFormat fmt = new DecimalFormat("#.##%");
fmt.parse("123.45"); // result : exception
```

GWT

```java
NumberFormat fmt = NumberFormat.getFormat("#.##");
fmt.parse("123.45%"); // result : 1.2345 --> WRONG

NumberFormat fmt = NumberFormat.getFormat("#.##%");
fmt.parse("123.45%"); // result : 123.45 --> WRONG

NumberFormat fmt = NumberFormat.getFormat("#.##%");
fmt.parse("123.45"); // result : exception
```
the correct approach here is to remove all the scale logic from the `parseNumber` method, and in the parse `parse(String, int[])` divide the result of the `parseNumber` by the multiplier, increment the postion if the multiplier is not `1`.

but this might be a breaking change for any code that is using number format with percent, so in order to preserve compatibility we can go with just dividing by the multiplier and leave the scale logic intact, and fix the case for the  
```java
NumberFormat fmt = NumberFormat.getFormat("#.##%");
fmt.parse("12.34%"); 
```
without breaking legacy code.

**Comment by vegegoku** [comment]
Gerrit patch can be found here
https://gwt-review.googlesource.com/c/gwt/+/20840

**Comment by aepshteyn** [comment]
Here is the workaround I'm using in my project:

https://github.com/aepshteyn/trsoftware.commons/blob/master/src/solutions/trsoftware/commons/client/text/CorrectedNumberFormat.java

Excerpt:
```
  public double parse(String text) throws NumberFormatException {
    return maybeCorrectParseResult(numberFormat.parse(text));
  }


  private double maybeCorrectParseResult(double parseResult) {
    return correctMultiplierError(parseResult);
  }


  /**
   * Attempts to correct the bug that {@link NumberFormat#parse(String)} fails to divide by the multiplier it
   * used in {@link NumberFormat#format(double)} (which happens when formatting percent, per mille, etc.)
   *
   * @param parseResult the value returned by {@link NumberFormat#parse(String)}
   * @return the result divided by the appropriate multiplier, if any
   * @see NumberFormat#multiplier
   * @see <a href="https://github.com/gwtproject/gwt/issues/9611">NumberFormat bug when parsing percentages</a>
   */
  private double correctMultiplierError(double parseResult) {
    // first, check that the multiplier error is actually present
    double multiplier = numberFormat.parse(numberFormat.format(1));
    if (!MathUtils.equal(1, multiplier, MathUtils.EPSILON)) {
      // the parse method probably failed to divide by the multiplier
      parseResult /= multiplier;
    }
    return parseResult;
  }
```
